File: domains/finance/projects/stock-market/development/software/ios/StockPrediction/VideoIntelligence/VideoIntelligenceManager.swift
import Foundation
import GRDB

// MARK: - ProcessingStatus

enum ProcessingStatus: Equatable {
    case queued
    case downloading
    case transcribing
    case extracting
    case done
    case failed(String)

    var displayLabel: String {
        switch self {
        case .queued:       return "Queued"
        case .downloading:  return "Downloading"
        case .transcribing: return "Transcribing"
        case .extracting:   return "Extracting Signals"
        case .done:         return "Done"
        case .failed(let m): return "Failed: \(m)"
        }
    }

    var isFinal: Bool {
        switch self {
        case .done, .failed: return true
        default: return false
        }
    }
}

// MARK: - ProcessingJob

struct ProcessingJob: Identifiable {
    let id: String          // same as videoId
    var videoId: String
    var status: ProcessingStatus
    var progress: Double
    var statusMessage: String
    var startedAt: Date
}

// MARK: - VideoIntelligenceManager

@MainActor
final class VideoIntelligenceManager: ObservableObject {
    static let shared = VideoIntelligenceManager()

    @Published var processingJobs: [String: ProcessingJob] = [:]    // videoId → job
    @Published var recentVideos: [VideoSourceRecord] = []
    @Published var recentSignals: [VideoSignalRecord] = []
    @Published var errorMessage: String? = nil

    private init() {
        Task { await loadRecent() }
    }

    // MARK: - Process YouTube URL

    /// End-to-end pipeline: download audio → transcribe → extract signals → persist.
    func processURL(_ url: String, modelOverride: String? = nil) async {
        let sanitized = url.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !sanitized.isEmpty else {
            errorMessage = "Please enter a YouTube URL."
            return
        }

        // Derive a provisional videoId for the job key before we have metadata
        let provisionalId = UUID().uuidString
        var job = ProcessingJob(
            id: provisionalId,
            videoId: provisionalId,
            status: .queued,
            progress: 0,
            statusMessage: "Starting…",
            startedAt: Date()
        )
        processingJobs[provisionalId] = job

        do {
            // 1. Download audio + fetch metadata
            job.status = .downloading
            job.statusMessage = "Fetching video info…"
            processingJobs[provisionalId] = job

            let (audioURL, metadata) = try await YouTubeAudioExtractor.shared.prepareAudio(
                youtubeURL: sanitized
            ) { progress, msg in
                Task { @MainActor [weak self] in
                    guard let self else { return }
                    var j = self.processingJobs[provisionalId] ?? job
                    j.progress = progress * 0.45
                    j.statusMessage = msg
                    self.processingJobs[provisionalId] = j
                }
            }

            // Persist video source record
            let videoRecord = VideoSourceRecord(
                id: metadata.videoId,
                url: sanitized,
                videoId: metadata.videoId,
                title: metadata.title,
                channelName: metadata.channelName,
                channelId: metadata.channelId,
                speakerName: "",
                publishedAt: metadata.publishedAt,
                durationSec: metadata.durationSec,
                viewCount: metadata.viewCount,
                status: "transcribing",
                errorMsg: nil,
                transcriptModel: nil,
                fullText: nil,
                createdAt: Date()
            )
            try? DatabaseManager.shared.saveVideoSource(videoRecord)

            // Remap job key to real videoId
            processingJobs.removeValue(forKey: provisionalId)
            job = ProcessingJob(
                id: metadata.videoId,
                videoId: metadata.videoId,
                status: .transcribing,
                progress: 0.45,
                statusMessage: "Starting transcription…",
                startedAt: job.startedAt
            )
            processingJobs[metadata.videoId] = job
            let realId = metadata.videoId

            // 2. Transcribe
            job.status = .transcribing
            processingJobs[realId] = job

            let result = try await VideoTranscriber.shared.transcribe(audioURL: audioURL) { progress, msg in
                Task { @MainActor [weak self] in
                    guard let self else { return }
                    var j = self.processingJobs[realId] ?? job
                    j.progress = 0.45 + progress * 0.35
                    j.statusMessage = msg
                    self.processingJobs[realId] = j
                }
            }

            // Clean up audio temp file
            try? FileManager.default.removeItem(at: audioURL)

            // Update DB with transcript
            try? DatabaseManager.shared.updateVideoSourceStatus(
                id: realId,
                status: "extracting",
                errorMsg: nil,
                fullText: result.fullText
            )

            // 3. Extract signals
            job.status = .extracting
            job.progress = 0.80
            job.statusMessage = "Extracting market signals…"
            processingJobs[realId] = job

            let signals = try await VideoSignalExtractor.shared.extractSignals(
                transcript: result.fullText,
                title: metadata.title,
                channel: metadata.channelName,
                videoId: realId
            )

            try? DatabaseManager.shared.saveVideoSignals(signals)

            // Update DB status to done
            try? DatabaseManager.shared.updateVideoSourceStatus(
                id: realId,
                status: "done",
                errorMsg: nil,
                fullText: result.fullText
            )

            // 4. Mark done
            job.status = .done
            job.progress = 1.0
            job.statusMessage = "Done — \(signals.count) signal(s) extracted"
            processingJobs[realId] = job

            await loadRecent()

            // Remove completed job after a short delay
            try? await Task.sleep(nanoseconds: 4_000_000_000)
            processingJobs.removeValue(forKey: realId)

        } catch {
            let errMsg = error.localizedDescription
            job.status = .failed(errMsg)
            job.statusMessage = errMsg
            processingJobs[provisionalId] = job
            errorMessage = errMsg

            // Clean up if we have a real video id
            // (best-effort — we may not have it)
        }
    }

    // MARK: - Load Recent

    func loadRecent() async {
        do {
            recentVideos = try DatabaseManager.shared.getVideoSources(limit: 20)
            recentSignals = try DatabaseManager.shared.getVideoSignals(ticker: nil, days: 30)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Query Signals

    func querySignals(ticker: String?, days: Int) async -> [VideoSignalRecord] {
        do {
            return try DatabaseManager.shared.getVideoSignals(ticker: ticker, days: days)
        } catch {
            errorMessage = error.localizedDescription
            return []
        }
    }

    // MARK: - Apply Signal to Prediction

    /// Saves a video-derived signal record to the local database for use in predictions.
    func applySignalToPrediction(_ signal: VideoSignalRecord) async {
        // Persist the applied signal — future prediction engine will read from video_signals
        // filtered by extractedAt to incorporate it automatically.
        // For now we re-save (upsert-like behaviour) with updated extractedAt to
        // float it to the top of signal priority.
        var updated = signal
        updated.extractedAt = Date()
        do {
            try DatabaseManager.shared.saveVideoSignals([updated])
            await loadRecent()
        } catch {
            errorMessage = "Failed to apply signal: \(error.localizedDescription)"
        }
    }

    // MARK: - Channel Tracking

    func trackChannel(_ channelId: String, name: String, speaker: String) async {
        let record = ChannelTrackRecord(
            channelId: channelId,
            channelName: name,
            speakerName: speaker,
            autoProcess: true,
            timeRangeYears: 5,
            createdAt: Date()
        )
        do {
            try DatabaseManager.shared.saveChannelTrack(record)
        } catch {
            errorMessage = "Failed to save channel: \(error.localizedDescription)"
        }
    }

    func getTrackedChannels() async -> [ChannelTrackRecord] {
        do {
            return try DatabaseManager.shared.getChannelTracks()
        } catch {
            errorMessage = error.localizedDescription
            return []
        }
    }

    func removeTrackedChannel(_ channelId: String) async {
        do {
            try DatabaseManager.shared.removeChannelTrack(channelId: channelId)
        } catch {
            errorMessage = "Failed to remove channel: \(error.localizedDescription)"
        }
    }
}

File: domains/finance/projects/stock-market/development/software/android/core/src/main/java/com/prediction/stockmarket/videointelligence/VideoIntelligenceManager.kt
package com.prediction.stockmarket.videointelligence

import android.content.Context
import android.util.Log
import com.prediction.stockmarket.data.database.AppDatabase
import com.prediction.stockmarket.prediction.LLMInferenceEngine
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.Date
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

private const val TAG = "VideoIntelligenceManager"

data class ProcessingJob(
    val videoId: String,
    val url: String,
    var status: JobStatus,
    var progress: Float,
    var statusMessage: String,
    val startedAt: Long = System.currentTimeMillis()
)

enum class JobStatus { QUEUED, DOWNLOADING, TRANSCRIBING, EXTRACTING, DONE, FAILED }

/**
 * Pre-seeded list of influential financial speakers with their YouTube channel IDs.
 * Users can add/remove from this list in the app.
 */
val INFLUENTIAL_SPEAKERS = listOf(
    "UCEb7pOMZ5h3MdqN7fYPD5mw" to "Elon Musk",
    "UCIRYBXDze5krPDzAEOxFGVA" to "Warren Buffett",
    "UCTk_-XgFDfSf6EGpJQJhRdA" to "Jerome Powell",
    "UCeeFfhMcJa1kjtfZAGskOCA" to "Jensen Huang",
    "UCE_M8A5yxnLfW0KghEeajjw" to "Tim Cook",
    "UCimKczFRuRUKPPHarX6v_jQ" to "Cathie Wood",
    "UCFiJ9iqkIEYQlFa_WJyRGqQ" to "Jim Cramer",
    "UCVHVFAm24e5i3kEi0IG3ItQ" to "Michael Saylor",
)

@Singleton
class VideoIntelligenceManager @Inject constructor(
    private val context: Context,
    private val database: AppDatabase,
    private val okHttpClient: okhttp3.OkHttpClient,
    private val llmEngine: LLMInferenceEngine
) {
    private val audioExtractor = YouTubeAudioExtractor(okHttpClient)
    private val transcriber = VideoTranscriber(context)
    private val signalExtractor = VideoSignalExtractor(llmEngine)

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private val _jobs = MutableStateFlow<Map<String, ProcessingJob>>(emptyMap())
    val jobs: StateFlow<Map<String, ProcessingJob>> = _jobs.asStateFlow()

    // -------------------------------------------------------------------------
    // URL processing pipeline
    // -------------------------------------------------------------------------

    /**
     * Kicks off the full YVIS pipeline for a YouTube URL:
     * 1. Download audio
     * 2. Transcribe
     * 3. Extract market signals
     * 4. Persist to Room
     */
    fun processUrl(url: String) {
        val videoId = audioExtractor.extractVideoId(url) ?: run {
            Log.w(TAG, "Cannot extract video ID from URL: $url")
            return
        }

        if (_jobs.value.containsKey(videoId)) {
            val existing = _jobs.value[videoId]
            if (existing?.status in listOf(JobStatus.QUEUED, JobStatus.DOWNLOADING, JobStatus.TRANSCRIBING, JobStatus.EXTRACTING)) {
                Log.i(TAG, "Job for $videoId already in progress — skipping")
                return
            }
        }

        val job = ProcessingJob(
            videoId = videoId,
            url = url,
            status = JobStatus.QUEUED,
            progress = 0f,
            statusMessage = "Queued"
        )
        updateJob(videoId, job)

        scope.launch {
            runPipeline(url, videoId)
        }
    }

    private suspend fun runPipeline(url: String, videoId: String) {
        val cacheDir = context.cacheDir

        try {
            // Phase 1: Download audio
            updateJob(videoId) { it.copy(status = JobStatus.DOWNLOADING, statusMessage = "Fetching video…") }

            val (audioFile, metadata) = audioExtractor.prepareAudio(url, cacheDir) { progress, msg ->
                updateJob(videoId) { j ->
                    j.copy(progress = progress * 0.4f, statusMessage = msg)
                }
            }

            // Persist initial VideoSource record
            val sourceId = UUID.randomUUID().toString()
            database.videoSourceDao().upsert(
                VideoSourceEntity(
                    id = sourceId,
                    url = url,
                    videoId = videoId,
                    title = metadata.title,
                    channelName = metadata.channelName,
                    channelId = metadata.channelId,
                    speakerName = "",
                    publishedAt = metadata.publishedAt,
                    durationSec = metadata.durationSec,
                    viewCount = metadata.viewCount,
                    status = "transcribing",
                    errorMsg = null,
                    transcriptModel = null,
                    fullText = null,
                    createdAt = Date()
                )
            )

            // Phase 2: Transcribe
            updateJob(videoId) { it.copy(status = JobStatus.TRANSCRIBING, progress = 0.4f, statusMessage = "Transcribing audio…") }

            val transcriptionResult = transcriber.transcribe(audioFile) { progress, msg ->
                updateJob(videoId) { j ->
                    j.copy(progress = 0.4f + progress * 0.4f, statusMessage = msg)
                }
            }

            // Update source with transcript
            database.videoSourceDao().updateStatus(
                id = sourceId,
                status = "extracting",
                errorMsg = null,
                fullText = transcriptionResult.fullText
            )

            // Phase 3: Extract signals
            updateJob(videoId) { it.copy(status = JobStatus.EXTRACTING, progress = 0.8f, statusMessage = "Extracting market signals…") }

            val signals = signalExtractor.extractSignals(
                transcript = transcriptionResult.fullText,
                title = metadata.title,
                channel = metadata.channelName,
                videoId = sourceId
            )

            if (signals.isNotEmpty()) {
                database.videoSignalDao().insertAll(signals)
            }

            // Finalize
            database.videoSourceDao().updateStatus(
                id = sourceId,
                status = "done",
                errorMsg = null,
                fullText = transcriptionResult.fullText
            )

            updateJob(videoId) {
                it.copy(
                    status = JobStatus.DONE,
                    progress = 1f,
                    statusMessage = "Done — ${signals.size} signals extracted"
                )
            }

            Log.i(TAG, "Pipeline complete for $videoId: ${signals.size} signals")

            // Clean up cached audio file to free space
            audioFile.delete()

        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            Log.e(TAG, "Pipeline failed for $videoId: ${e.message}", e)
            updateJob(videoId) { it.copy(status = JobStatus.FAILED, statusMessage = "Error: ${e.message}") }
        }
    }

    // -------------------------------------------------------------------------
    // Query methods
    // -------------------------------------------------------------------------

    suspend fun querySignals(ticker: String?, days: Int): List<VideoSignalEntity> {
        val cutoff = Date(System.currentTimeMillis() - days.toLong() * 24 * 60 * 60 * 1000)
        return database.videoSignalDao().querySignals(
            ticker = ticker?.takeIf { it.isNotBlank() },
            cutoff = cutoff
        )
    }

    suspend fun applySignal(signal: VideoSignalEntity) {
        // Mark signal as applied — in a full implementation this would update
        // the PredictionEngine with the extracted weight/direction for the ticker.
        Log.i(TAG, "Applying signal: ticker=${signal.ticker}, dir=${signal.direction}, weight=${signal.weight}")
        // Future: database.predictionDao().adjustWithSignal(signal)
    }

    suspend fun getRecentVideos(limit: Int = 20): List<VideoSourceEntity> =
        database.videoSourceDao().getRecent(limit)

    // -------------------------------------------------------------------------
    // Channel tracking
    // -------------------------------------------------------------------------

    suspend fun trackChannel(channelId: String, name: String, speaker: String) {
        database.channelTrackDao().upsert(
            ChannelTrackEntity(
                channelId = channelId,
                channelName = name,
                speakerName = speaker,
                autoProcess = true,
                timeRangeYears = 1,
                createdAt = Date()
            )
        )
        Log.i(TAG, "Tracking channel: $name ($channelId)")
    }

    suspend fun getTrackedChannels(): List<ChannelTrackEntity> =
        database.channelTrackDao().getAll()

    suspend fun removeTrackedChannel(channelId: String) {
        database.channelTrackDao().remove(channelId)
        Log.i(TAG, "Removed channel tracking: $channelId")
    }

    // -------------------------------------------------------------------------
    // Whisper model info
    // -------------------------------------------------------------------------

    fun getWhisperModels(): List<WhisperModelInfo> = WHISPER_MODELS.map { m ->
        m  // Downloaded status is checked via VideoTranscriber.isModelDownloaded()
    }

    fun isWhisperModelDownloaded(modelId: String): Boolean =
        transcriber.isModelDownloaded(modelId)

    suspend fun downloadWhisperModel(
        modelId: String,
        onProgress: (Float, String) -> Unit
    ) = transcriber.downloadModel(modelId, onProgress)

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    private fun updateJob(videoId: String, update: (ProcessingJob) -> ProcessingJob) {
        val current = _jobs.value.toMutableMap()
        val existing = current[videoId] ?: return
        current[videoId] = update(existing)
        _jobs.value = current
    }

    private fun updateJob(videoId: String, job: ProcessingJob) {
        val current = _jobs.value.toMutableMap()
        current[videoId] = job
        _jobs.value = current
    }
}
